### PR TITLE
[12.0] [IMP] geoengine_swisstopo : Add Swisstopo projections

### PR DESCRIPTION
**Pull request by sbrunner**

### CODE REVIEW

**Review by cyrilmanuel — APPROVED** [review]
for the code, LGTM :) 

 but for the 3 commits, i'm not sure if you shall split them in 2 (pre-commit and swisstopo projections) 

same here for the name of the PR. i'm not sure if shall be rename to contain the version, type of pr (implementation) and a title explicit about the module that contain most of the modification and a description of the implementation. example  : 
[12.0] [IMP] base_geoengine : Adding Swisstopo projections 

@yvaucher could you confirm or refute this :))

File: geoengine_swisstopo/static/src/js/geoengine_swisstopo.js [hunk 1]
```diff
@@ -1,124 +1,225 @@
...
+            init_proj4(this);
+        },
+        _render: function (parent) {
+            define_projections();
+            this._super.apply(this, arguments);
+        },
+
+    });
+});
+odoo.define("geoengine_swisstopo.projection_EPSG_2056", function (require) {
```
Comment (yvaucher) [inline geoengine_swisstopo/static/src/js/geoengine_swisstopo.js]: Seems unecessary to me to call `init_proj4` per projection.

File: geoengine_swisstopo/static/src/js/geoengine_swisstopo.js [hunk 2]
```diff
@@ -1,124 +1,225 @@
...
+            this._super.apply(this, arguments);
+            init_proj4(this);
+        },
+        _render: function (parent) {
+            define_projections();
+            this._super.apply(this, arguments);
+        },
+    });
+});
+odoo.define("geoengine_swisstopo.projection_EPSG_21781", function (require) {
```
Comment (yvaucher) [inline geoengine_swisstopo/static/src/js/geoengine_swisstopo.js]: To rename `geoengine_swisstopo.init_projections` ? And have a single call to define_projections and init_proj4.

File: geoengine_swisstopo/static/src/js/geoengine_swisstopo.js [hunk 3]
```diff
@@ -1,124 +1,225 @@
...
 
-/**
- * Extents of Swiss projections. (EPSG:21781)
- */
-var EXTENT = [420000, 30000, 900000, 350000];
+odoo.define("geoengine_swisstopo.projection_EPSG_4326", function (require) {
+    "use strict";
 
-var PROJECTION_CODE = "EPSG:21781";
+    const GeoengineWidgets = require("base_geoengine.geoengine_widgets");
```
Comment (yvaucher) [inline geoengine_swisstopo/static/src/js/geoengine_swisstopo.js:72]: The `require` is similiar as an `import` wouldn't it make more sense to keep it at the root level of the file or was there some cyclic import issue?

File: geoengine_swisstopo/geo_view/geo_raster_layer.py [hunk 4]
```diff
@@ -1,7 +1,15 @@
...
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from odoo import fields, models
+import logging
+
+import requests
+from odoo import api, fields, models
+
+_LOGGER = logging.getLogger(__name__)
+_CAPABILITIES_URL = (
```
Comment (yvaucher) [inline geoengine_swisstopo/geo_view/geo_raster_layer.py:11]: As those 2 variables are defined in the last commit, it seems the splitting of commit is odd.

**Review by yvaucher — CHANGES_REQUESTED** [review]
Looks great, just have a little concern on the multiple odoo.define for projections as it seems redundant.

Plus about the commits, some part of the changes of the 2nd commit deserve to be in the 3rd commit (or be squashed) the linting could go in a separate commit to let your last commit emphasis on the practical changes.

File: geoengine_swisstopo/geo_view/geo_raster_layer.py [hunk 5]
```diff
@@ -5,8 +5,53 @@
...
+                    _LOGGER.error(
+                        "Swisstopo WMTS Capabilities request (%s)\n"
+                        "failed with status code %s:\n%s",
+                        url,
+                        response.status_code,
+                        response.text,
+                    )
+                    record.capabilities = False
+            else:
+                record.capabilities = False
```
Comment (yvaucher) [inline geoengine_swisstopo/geo_view/geo_raster_layer.py]: This doesn't looks like it should be part of the pre-commit

Comment (sbrunner) [reply geoengine_swisstopo/geo_view/geo_raster_layer.py]: Done, tanks :-)

File: geoengine_swisstopo/static/src/js/geoengine_swisstopo.js (hunk 3 above)
Comment (sbrunner) [reply geoengine_swisstopo/static/src/js/geoengine_swisstopo.js:72]: I just keep it as it was, and the require function is not defined at top level...

File: geoengine_swisstopo/geo_view/geo_raster_layer.py (hunk 4 above)
Comment (sbrunner) [reply geoengine_swisstopo/geo_view/geo_raster_layer.py:11]: Looks to be fixed in the last version :-)

**Review by yvaucher — APPROVED** [review]

**Review by cyrilmanuel — APPROVED** [review]

File: base_geoengine/geo_model.py [hunk 6]
```diff
@@ -131,7 +131,7 @@ def get_edit_info_for_geo_column(self, column):
             'default_extent': view.default_extent or DEFAULT_EXTENT,
             'default_zoom': view.default_zoom,
         }
-        logger.debug("Parameters for geo field {}:\n{}".format(column, res))
+        logger.debug(f"Parameters for geo field {column}:\n{res}")
```
Comment (HekkiMelody) [inline base_geoengine/geo_model.py:134]: the minimum version of python supported by v12 is 3.5, which does not have fstrings. This unfortunately broke my installation.

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -56,4 +56,45 @@ repos:
...
+          - --py36-plus
+  # - repo: https://github.com/PyCQA/isort
+  #   rev: 5.12.0
+  #   language_version: python3.8
+  #   hooks:
+  #     - id: isort
+  #       name: isort except __init__.py
+  #       args:
+  #         - --settings=.
+  #       exclude: /__init__\.py$
```
Comment (HekkiMelody) [inline .pre-commit-config.yaml:100]: This shouldn't be modified manually, but only through `copier update`

**Comment by HekkiMelody** [comment]
Fixed the issue in https://github.com/OCA/geospatial/pull/366

**Pull request merged**